File: packages/wasmvm/wasmlib/go/wasmlib/sandbox.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

import (
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmrequests"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

const (
	FnAccountID           = int32(-1)
	FnAllowance           = int32(-2)
	FnBalance             = int32(-3)
	FnBalances            = int32(-4)
	FnBlockContext        = int32(-5)
	FnCall                = int32(-6)
	FnCaller              = int32(-7)
	FnChainID             = int32(-8)
	FnChainOwnerID        = int32(-9)
	FnContract            = int32(-10)
	FnDeployContract      = int32(-11)
	FnEntropy             = int32(-12)
	FnEstimateDust        = int32(-13)
	FnEvent               = int32(-14)
	FnLog                 = int32(-15)
	FnMinted              = int32(-16)
	FnPanic               = int32(-17)
	FnParams              = int32(-18)
	FnPost                = int32(-19)
	FnRequest             = int32(-20)
	FnRequestID           = int32(-21)
	FnResults             = int32(-22)
	FnSend                = int32(-23)
	FnStateAnchor         = int32(-24)
	FnTimestamp           = int32(-25)
	FnTrace               = int32(-26)
	FnTransferAllowed     = int32(-27)
	FnUtilsBech32Decode   = int32(-28)
	FnUtilsBech32Encode   = int32(-29)
	FnUtilsBlsAddress     = int32(-30)
	FnUtilsBlsAggregate   = int32(-31)
	FnUtilsBlsValid       = int32(-32)
	FnUtilsEd25519Address = int32(-33)
	FnUtilsEd25519Valid   = int32(-34)
	FnUtilsHashBlake2b    = int32(-35)
	FnUtilsHashName       = int32(-36)
	FnUtilsHashSha3       = int32(-37)
)

type ScSandbox struct{}

// TODO go over core contract schemas to set correct unsigned types

func Log(text string) {
	Sandbox(FnLog, []byte(text))
}

func Panic(text string) {
	Sandbox(FnPanic, []byte(text))
}

func Trace(text string) {
	Sandbox(FnTrace, []byte(text))
}

func NewParamsProxy() wasmtypes.Proxy {
	return wasmtypes.NewProxy(NewScDictFromBytes(Sandbox(FnParams, nil)))
}

// retrieve the agent id of this contract account
func (s ScSandbox) AccountID() wasmtypes.ScAgentID {
	return wasmtypes.AgentIDFromBytes(Sandbox(FnAccountID, nil))
}

func (s ScSandbox) Balance(tokenID wasmtypes.ScTokenID) uint64 {
	return wasmtypes.Uint64FromBytes(Sandbox(FnBalance, tokenID.Bytes()))
}

// access the current balances for all assets
func (s ScSandbox) Balances() *ScBalances {
	balances := NewScAssets(Sandbox(FnBalances, nil)).Balances()
	return &balances
}

// calls a smart contract function
func (s ScSandbox) callWithAllowance(hContract, hFunction wasmtypes.ScHname, params *ScDict, allowance *ScTransfer) *ScImmutableDict {
	req := &wasmrequests.CallRequest{
		Contract:  hContract,
		Function:  hFunction,
		Params:    params.Bytes(),
		Allowance: allowance.Bytes(),
	}
	res := Sandbox(FnCall, req.Bytes())
	return NewScDictFromBytes(res).Immutable()
}

// retrieve the agent id of the owner of the chain this contract lives on
func (s ScSandbox) ChainOwnerID() wasmtypes.ScAgentID {
	return wasmtypes.AgentIDFromBytes(Sandbox(FnChainOwnerID, nil))
}

// retrieve the hname of this contract
func (s ScSandbox) Contract() wasmtypes.ScHname {
	return wasmtypes.HnameFromBytes(Sandbox(FnContract, nil))
}

// retrieve the chain id of the chain this contract lives on
func (s ScSandbox) CurrentChainID() wasmtypes.ScChainID {
	return wasmtypes.ChainIDFromBytes(Sandbox(FnChainID, nil))
}

// logs informational text message
func (s ScSandbox) Log(text string) {
	Sandbox(FnLog, []byte(text))
}

// logs error text message and then panics
func (s ScSandbox) Panic(text string) {
	Sandbox(FnPanic, []byte(text))
}

// retrieve parameters passed to the smart contract function that was called
func (s ScSandbox) Params() *ScImmutableDict {
	return NewScDictFromBytes(Sandbox(FnParams, nil)).Immutable()
}

func (s ScSandbox) RawState() ScImmutableState {
	return ScImmutableState{}
}

// panics if condition is not satisfied
func (s ScSandbox) Require(cond bool, msg string) {
	if !cond {
		s.Panic(msg)
	}
}

func (s ScSandbox) Results(results *ScDict) {
	Sandbox(FnResults, results.Bytes())
}

// deterministic timestamp in nanosecond fixed at the moment of calling the smart contract
func (s ScSandbox) Timestamp() uint64 {
	return wasmtypes.Uint64FromBytes(Sandbox(FnTimestamp, nil))
}

// logs debugging trace text message
func (s ScSandbox) Trace(text string) {
	Sandbox(FnTrace, []byte(text))
}

// access diverse utility functions
func (s ScSandbox) Utility() ScSandboxUtils {
	return ScSandboxUtils{}
}

type ScSandboxView struct {
	ScSandbox
}

// calls a smart contract view
func (s ScSandboxView) Call(hContract, hFunction wasmtypes.ScHname, params *ScDict) *ScImmutableDict {
	return s.callWithAllowance(hContract, hFunction, params, nil)
}

type ScSandboxFunc struct {
	ScSandbox
}

// access the allowance assets
func (s ScSandboxFunc) Allowance() *ScBalances {
	buf := Sandbox(FnAllowance, nil)
	balances := NewScAssets(buf).Balances()
	return &balances
}

//func (s ScSandbox) BlockContext(construct func(sandbox ScSandbox) interface{}, onClose func(interface{})) interface{} {
//	panic("implement me")
//}

// calls a smart contract function
func (s ScSandboxFunc) Call(hContract, hFunction wasmtypes.ScHname, params *ScDict, allowance *ScTransfer) *ScImmutableDict {
	return s.callWithAllowance(hContract, hFunction, params, allowance)
}

// retrieve the agent id of the caller of the smart contract
func (s ScSandboxFunc) Caller() wasmtypes.ScAgentID {
	return wasmtypes.AgentIDFromBytes(Sandbox(FnCaller, nil))
}

// deploys a smart contract
func (s ScSandboxFunc) DeployContract(programHash wasmtypes.ScHash, name, description string, initParams *ScDict) {
	req := &wasmrequests.DeployRequest{
		ProgHash:    programHash,
		Name:        name,
		Description: description,
		Params:      initParams.Bytes(),
	}
	Sandbox(FnDeployContract, req.Bytes())
}

// returns random entropy data for current request.
func (s ScSandboxFunc) Entropy() wasmtypes.ScHash {
	return wasmtypes.HashFromBytes(Sandbox(FnEntropy, nil))
}

func (s ScSandboxFunc) EstimateDust(fn *ScFunc) uint64 {
	req := &wasmrequests.PostRequest{
		Contract:  fn.hContract,
		Function:  fn.hFunction,
		Params:    fn.params.Bytes(),
		Allowance: fn.allowance.Bytes(),
		Transfer:  fn.transfer.Bytes(),
		Delay:     fn.delay,
	}
	return wasmtypes.Uint64FromBytes(Sandbox(FnEstimateDust, req.Bytes()))
}

// signals an event on the node that external entities can subscribe to
func (s ScSandboxFunc) Event(msg string) {
	Sandbox(FnEvent, []byte(msg))
}

// retrieve the assets that were minted in this transaction
func (s ScSandboxFunc) Minted() ScBalances {
	return NewScAssets(Sandbox(FnMinted, nil)).Balances()
}

// Post (delayed) posts a SC function request
func (s ScSandboxFunc) Post(chainID wasmtypes.ScChainID, hContract, hFunction wasmtypes.ScHname, params *ScDict, allowance, transfer ScTransfer, delay uint32) {
	req := &wasmrequests.PostRequest{
		ChainID:   chainID,
		Contract:  hContract,
		Function:  hFunction,
		Params:    params.Bytes(),
		Allowance: allowance.Bytes(),
		Transfer:  transfer.Bytes(),
		Delay:     delay,
	}
	Sandbox(FnPost, req.Bytes())
}

var (
	entropy []byte
	offset  = 0
)

// generates a random value from 0 to max (exclusive max) using a deterministic RNG
func (s ScSandboxFunc) Random(max uint64) (rnd uint64) {
	if max == 0 {
		s.Panic("random: max parameter should be non-zero")
	}

	// note that entropy gets reset for every request
	if len(entropy) == 0 {
		// first time in this request, initialize with current request entropy
		entropy = s.Entropy().Bytes()
		offset = 0
	}
	if offset == 32 {
		// ran out of entropy data, hash entropy for next pseudo-random entropy
		entropy = s.Utility().HashBlake2b(entropy).Bytes()
		offset = 0
	}
	rnd = wasmtypes.Uint64FromBytes(entropy[offset:offset+8]) % max
	offset += 8
	return rnd
}

func (s ScSandboxFunc) RawState() ScState {
	return ScState{}
}

//func (s ScSandboxFunc) Request() ScRequest {
//	panic("implement me")
//}

// retrieve the request id of this transaction
func (s ScSandboxFunc) RequestID() wasmtypes.ScRequestID {
	return wasmtypes.RequestIDFromBytes(Sandbox(FnRequestID, nil))
}

// Send transfers SC assets to the specified address
func (s ScSandboxFunc) Send(address wasmtypes.ScAddress, transfer *ScTransfer) {
	// we need some assets to send
	if transfer.IsEmpty() {
		return
	}

	req := wasmrequests.SendRequest{
		Address:  address,
		Transfer: transfer.Bytes(),
	}
	Sandbox(FnSend, req.Bytes())
}

//func (s ScSandboxFunc) StateAnchor() interface{} {
//	panic("implement me")
//}

// TransferAllowed transfers allowed assets from caller to the specified account
func (s ScSandboxFunc) TransferAllowed(agentID wasmtypes.ScAgentID, transfer *ScTransfer, create bool) {
	// we need some assets to send
	if transfer.IsEmpty() {
		return
	}

	req := wasmrequests.TransferRequest{
		AgentID:  agentID,
		Create:   create,
		Transfer: transfer.Bytes(),
	}
	Sandbox(FnTransferAllowed, req.Bytes())
}

File: packages/wasmvm/wasmlib/ts/wasmlib/sandbox.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmrequests from "./wasmrequests"
import * as wasmtypes from "./wasmtypes"
import {ScAssets, ScBalances, ScTransfer} from "./assets";
import {ScDict, ScImmutableDict} from "./dict";
import {sandbox} from "./host";
import {ScSandboxUtils} from "./sandboxutils";
import {ScImmutableState, ScState} from "./state";
import {ScFunc} from "./contract";

// @formatter:off
export const FnAccountID           : i32 = -1;
export const FnAllowance           : i32 = -2;
export const FnBalance             : i32 = -3;
export const FnBalances            : i32 = -4;
export const FnBlockContext        : i32 = -5;
export const FnCall                : i32 = -6;
export const FnCaller              : i32 = -7;
export const FnChainID             : i32 = -8;
export const FnChainOwnerID        : i32 = -9;
export const FnContract            : i32 = -10;
export const FnDeployContract      : i32 = -12;
export const FnEntropy             : i32 = -13;
export const FnEstimateDust        : i32 = -14;
export const FnEvent               : i32 = -15;
export const FnLog                 : i32 = -16;
export const FnMinted              : i32 = -17;
export const FnPanic               : i32 = -18;
export const FnParams              : i32 = -19;
export const FnPost                : i32 = -20;
export const FnRequest             : i32 = -21;
export const FnRequestID           : i32 = -22;
export const FnResults             : i32 = -23;
export const FnSend                : i32 = -24;
export const FnStateAnchor         : i32 = -25;
export const FnTimestamp           : i32 = -26;
export const FnTrace               : i32 = -27;
export const FnTransferAllowed     : i32 = -28;
export const FnUtilsBase58Decode   : i32 = -29;
export const FnUtilsBase58Encode   : i32 = -30;
export const FnUtilsBech32Decode   : i32 = -31;
export const FnUtilsBech32Encode   : i32 = -32;
export const FnUtilsBlsAddress     : i32 = -33;
export const FnUtilsBlsAggregate   : i32 = -34;
export const FnUtilsBlsValid       : i32 = -35;
export const FnUtilsEd25519Address : i32 = -36;
export const FnUtilsEd25519Valid   : i32 = -37;
export const FnUtilsHashBlake2b    : i32 = -38;
export const FnUtilsHashName       : i32 = -39;
export const FnUtilsHashSha3       : i32 = -40;
// @formatter:on

// Direct logging of text to host log
export function log(text: string): void {
    sandbox(FnLog, wasmtypes.stringToBytes(text));
}

// Direct logging of error to host log, followed by panicking out of the Wasm code
export function panic(text: string): void {
    sandbox(FnPanic, wasmtypes.stringToBytes(text));
}

// Direct conditional logging of debug-level informational text to host log
export function trace(text: string): void {
    sandbox(FnTrace, wasmtypes.stringToBytes(text));
}

export function paramsProxy(): wasmtypes.Proxy {
    return new ScDict(sandbox(FnParams, null)).asProxy();
}

export class ScSandbox {
    // retrieve the agent id of this contract account
    public accountID(): wasmtypes.ScAgentID {
        return wasmtypes.agentIDFromBytes(sandbox(FnAccountID, null));
    }

    public balance(tokenID: wasmtypes.ScTokenID): u64 {
        return wasmtypes.uint64FromBytes(sandbox(FnBalance, tokenID.toBytes()));
    }

    // access the current balances for all assets
    public balances(): ScBalances {
        return new ScAssets(sandbox(FnBalances, null)).balances();
    }

    // calls a smart contract function
    protected callWithAllowance(hContract: wasmtypes.ScHname, hFunction: wasmtypes.ScHname, params: ScDict | null, allowance: ScTransfer | null): ScImmutableDict {
        const req = new wasmrequests.CallRequest();
        req.contract = hContract;
        req.function = hFunction;
        if (params === null) {
            params = new ScDict([]);
        }
        req.params = params.toBytes();
        if (allowance === null) {
            allowance = new ScTransfer();
        }
        req.allowance = allowance.toBytes();
        const res = sandbox(FnCall, req.bytes());
        return new ScDict(res).immutable();
    }

    // retrieve the agent id of the owner of the chain this contract lives on
    public chainOwnerID(): wasmtypes.ScAgentID {
        return wasmtypes.agentIDFromBytes(sandbox(FnChainOwnerID, null));
    }

    // retrieve the hname of this contract
    public contract(): wasmtypes.ScHname {
        return wasmtypes.hnameFromBytes(sandbox(FnContract, null));
    }

    // retrieve the chain id of the chain this contract lives on
    public currentChainID(): wasmtypes.ScChainID {
        return wasmtypes.chainIDFromBytes(sandbox(FnChainID, null));
    }

    // logs informational text message
    public log(text: string): void {
        sandbox(FnLog, wasmtypes.stringToBytes(text));
    }

    // logs error text message and then panics
    public panic(text: string): void {
        sandbox(FnPanic, wasmtypes.stringToBytes(text));
    }

    // retrieve parameters passed to the smart contract function that was called
    public params(): ScImmutableDict {
        return new ScDict(sandbox(FnParams, null)).immutable();
    }

    // panics if condition is not satisfied
    public require(cond: bool, msg: string): void {
        if (!cond) {
            this.panic(msg);
        }
    }

    public results(results: ScDict): void {
        sandbox(FnResults, results.toBytes());
    }

    // deterministic time stamp fixed at the moment of calling the smart contract
    public timestamp(): u64 {
        return wasmtypes.uint64FromBytes(sandbox(FnTimestamp, null));
    }

    // logs debugging trace text message
    public trace(text: string): void {
        sandbox(FnTrace, wasmtypes.stringToBytes(text));
    }

    // access diverse utility functions
    public utility(): ScSandboxUtils {
        return new ScSandboxUtils();
    }
}

export class ScSandboxView extends ScSandbox {
    // calls a smart contract view
    public call(hContract: wasmtypes.ScHname, hFunction: wasmtypes.ScHname, params: ScDict | null): ScImmutableDict {
        return this.callWithAllowance(hContract, hFunction, params, null);
    }

    public rawState(): ScImmutableState {
        return new ScImmutableState();
    }
}

export class ScSandboxFunc extends ScSandbox {
    private static entropy: u8[] = [];
    private static offset: u32 = 0;

    // access the allowance assets
    public allowance(): ScBalances {
        const buf = sandbox(FnAllowance, null);
        return new ScAssets(buf).balances();
    }

    //public blockContext(construct func(sandbox: ScSandbox) interface{}, onClose func(interface{})): interface{} {
    //	panic("implement me")
    //}

    // calls a smart contract function
    public call(hContract: wasmtypes.ScHname, hFunction: wasmtypes.ScHname, params: ScDict | null, allowance: ScTransfer | null): ScImmutableDict {
        return this.callWithAllowance(hContract, hFunction, params, allowance);
    }

    // retrieve the agent id of the caller of the smart contract
    public caller(): wasmtypes.ScAgentID {
        return wasmtypes.agentIDFromBytes(sandbox(FnCaller, null));
    }

    // deploys a smart contract
    public deployContract(programHash: wasmtypes.ScHash, name: string, description: string, initParams: ScDict | null): void {
        if (initParams === null) {
            initParams = new ScDict([]);
        }
        const req = new wasmrequests.DeployRequest();
        req.progHash = programHash;
        req.name = name;
        req.description = description;
        req.params = initParams.toBytes();
        sandbox(FnDeployContract, req.bytes());
    }

    // returns random entropy data for current request.
    public entropy(): wasmtypes.ScHash {
        return wasmtypes.hashFromBytes(sandbox(FnEntropy, null));
    }

    public estimateDust(fn: ScFunc): u64 {
        const req = new wasmrequests.PostRequest();
        req.contract = fn.hContract;
        req.function = fn.hFunction;
        req.params = fn.params.toBytes();
        let allowance = fn.allowanceAssets;
        if (allowance === null) {
            allowance = new ScTransfer();
        }
        req.allowance = allowance.toBytes();
        let transfer = fn.transferAssets;
        if (transfer === null) {
            transfer = new ScTransfer();
        }
        req.transfer = transfer.toBytes();
        req.delay = fn.delaySeconds;
        return wasmtypes.uint64FromBytes(sandbox(FnEstimateDust, req.bytes()));
    }

    // signals an event on the node that external entities can subscribe to
    public event(msg: string): void {
        sandbox(FnEvent, wasmtypes.stringToBytes(msg));
    }

    // retrieve the assets that were minted in this transaction
    public minted(): ScBalances {
        return new ScAssets(sandbox(FnMinted, null)).balances();
    }

    // Post (delayed) posts a SC function request
    public post(chainID: wasmtypes.ScChainID, hContract: wasmtypes.ScHname, hFunction: wasmtypes.ScHname, params: ScDict, allowance: ScTransfer, transfer: ScTransfer, delay: u32): void {
        const req = new wasmrequests.PostRequest();
        req.chainID = chainID;
        req.contract = hContract;
        req.function = hFunction;
        req.params = params.toBytes();
        req.allowance = allowance.toBytes();
        req.transfer = transfer.toBytes();
        req.delay = delay;
        sandbox(FnPost, req.bytes());
    }

    // generates a random value from 0 to max (exclusive: max) using a deterministic RNG
    public random(max: u64): u64 {
        if (max == 0) {
            this.panic("random: max parameter should be non-zero");
        }

        // note that entropy gets reset for every request
        if (ScSandboxFunc.entropy.length == 0) {
            // first time in this: request, initialize with current request entropy
            ScSandboxFunc.entropy = this.entropy().toBytes();
            ScSandboxFunc.offset = 0;
        }
        if (ScSandboxFunc.offset == 32) {
            // ran out of entropy: data, hash entropy for next pseudo-random entropy
            ScSandboxFunc.entropy = this.utility().hashBlake2b(ScSandboxFunc.entropy).toBytes();
            ScSandboxFunc.offset = 0;
        }
        let rnd = wasmtypes.uint64FromBytes(ScSandboxFunc.entropy.slice(ScSandboxFunc.offset, ScSandboxFunc.offset + 8)) % max;
        ScSandboxFunc.offset += 8;
        return rnd;
    }

    public rawState(): ScState {
        return new ScState();
    }

    //public request(): ScRequest {
    //	panic("implement me")
    //}

    // retrieve the request id of this transaction
    public requestID(): wasmtypes.ScRequestID {
        return wasmtypes.requestIDFromBytes(sandbox(FnRequestID, null));
    }

    // Send transfers SC assets to the specified address
    public send(address: wasmtypes.ScAddress, transfer: ScTransfer): void {
        // we need some assets to send
        if (transfer.isEmpty()) {
            return;
        }

        const req = new wasmrequests.SendRequest();
        req.address = address;
        req.transfer = transfer.toBytes();
        sandbox(FnSend, req.bytes());
    }

    //public stateAnchor(): interface{} {
    //	panic("implement me")
    //}

    // TransferAllowed transfers allowed assets from caller to the specified account
    public transferAllowed(agentID: wasmtypes.ScAgentID, transfer: ScTransfer, create: bool): void {
        // we need some assets to send
        if (transfer.isEmpty()) {
            return;
        }

        const req = new wasmrequests.TransferRequest();
        req.agentID = agentID;
        req.create = create;
        req.transfer = transfer.toBytes();
        sandbox(FnTransferAllowed, req.bytes());
    }
}
